File: src/main/java/com/github/dkoval/leetcode/problems/PossibleBipartition.java
package com.github.dkoval.leetcode.problems;

import java.util.*;

/**
 * <a href="https://leetcode.com/problems/possible-bipartition/">Possible Bipartition</a>
 * <p>
 * We want to split a group of n people (labeled from 1 to n) into two groups of any size.
 * Each person may dislike some other people, and they should not go into the same group.
 * <p>
 * Given the integer n and the array dislikes where dislikes[i] = [ai, bi] indicates that the person labeled ai
 * does not like the person labeled bi, return true if it is possible to split everyone into two groups in this way.
 * <p>
 * Constraints:
 * <ul>
 *  <li>1 <= n <= 2000</li>
 *  <li>0 <= dislikes.length <= 10^4</li>
 *  <li>dislikes[i].length == 2</li>
 *  <li>1 <= dislikes[i][j] <= n</li>
 *  <li>ai < bi</li>
 *  <li>All the pairs of dislikes are unique</li>
 * </ul>
 */
public interface PossibleBipartition {

    boolean possibleBipartition(int n, int[][] dislikes);

    class PossibleBipartitionRev1 implements PossibleBipartition {

        @Override
        public boolean possibleBipartition(int n, int[][] dislikes) {
            Map<Integer, List<Integer>> graph = new HashMap<>();
            for (int[] people : dislikes) {
                graph.computeIfAbsent(people[0], key -> new ArrayList<>()).add(people[1]);
                graph.computeIfAbsent(people[1], key -> new ArrayList<>()).add(people[0]);
            }

            int[] colors = new int[n + 1];
            Arrays.fill(colors, -1);
            for (int u = 0; u < n; u++) {
                if (colors[u] == -1 && !dfs(graph, u,  0, colors)) {
                    return false;
                }
            }
            return true;
        }

        private boolean dfs(Map<Integer, List<Integer>> graph, int u, int color, int[] colors) {
            colors[u] = color;
            int adjColor = 1 - color;
            for (int v : graph.getOrDefault(u, Collections.emptyList())) {
                if (colors[v] == -1) {
                    boolean ok = dfs(graph, v, adjColor, colors);
                    if (!ok) {
                        return false;
                    }
                } else if (colors[v] != adjColor) {
                    return false;
                }
            }
            return true;
        }
    }
}

File: src/main/kotlin/com/github/dkoval/leetcode/challenge/PossibleBipartition.kt
package com.github.dkoval.leetcode.challenge

/**
 * [Possible Bipartition](https://leetcode.com/explore/challenge/card/may-leetcoding-challenge/537/week-4-may-22nd-may-28th/3342/)
 *
 * Given a set of N people (numbered 1, 2, ..., N), we would like to split everyone into two groups of any size.
 * Each person may dislike some other people, and they should not go into the same group.
 * Formally, if ```dislikes[i] = [a, b]```, it means it is not allowed to put the people numbered a and b into the same group.
 *
 * Return true if and only if it is possible to split everyone into two groups in this way.
 */
object PossibleBipartition {

    fun possibleBipartition(N: Int, dislikes: Array<IntArray>): Boolean {
        val adj = adjListOf(N, dislikes)
        val groups = Array<Group?>(N) { null }
        for (i in 0 until N) {
            if (groups[i] == null && !dfs(i, Group.ZERO, adj, groups)) {
                return false
            }
        }
        return true
    }

    private fun adjListOf(N: Int, dislikes: Array<IntArray>): Array<MutableList<Int>> {
        val adj = Array(N) { mutableListOf<Int>() }
        for ((a, b) in dislikes) {
            adj[a - 1].add(b - 1)
            adj[b - 1].add(a - 1)
        }
        return adj
    }

    private fun dfs(v: Int, group: Group, adj: Array<MutableList<Int>>, groups: Array<Group?>): Boolean {
        if (groups[v] == null) {
            groups[v] = group
        } else {
            return groups[v] == group
        }
        for (u in adj[v]) {
            if (!dfs(u, group.flip(), adj, groups)) {
                return false
            }
        }
        return true
    }

    private enum class Group {
        ZERO {
            override fun flip(): Group = ONE
        },
        ONE {
            override fun flip(): Group = ZERO
        };

        abstract fun flip(): Group
    }
}